### ISSUE TITLE
Incorrect execution of the wireless module with initial skipped bytes

### ISSUE THREAD

**Issue by whymarrh** [LakeMaps/microcontrollers#12]
The following execution was seen during testing:

```
Asking for message
Send
Recv
Read message
Read state is microcontrollers.MessageState$Started@1f93396
I'm skipping -1
Read state is microcontrollers.MessageState$Started@d1248f
I'm skipping 0
Read state is microcontrollers.MessageState$Started@74c380
I'm skipping 0
Read state is microcontrollers.MessageState$Started@11c2b83
I'm skipping 0
Read state is microcontrollers.MessageState$Started@bedd29
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1ca7dad
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1cccf97
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1b2a89e
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1597954
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1e64f07
I'm skipping 0
Read state is microcontrollers.MessageState$Started@959f1b
I'm skipping 0
Read state is microcontrollers.MessageState$Started@6d3e4f
I'm skipping 0
Read state is microcontrollers.MessageState$Started@127a8f
I'm skipping 0
Read state is microcontrollers.MessageState$Started@101803d
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1bca801
I'm skipping 0
Read state is microcontrollers.MessageState$Started@38697f
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1f3b6f5
I'm skipping 0
Read state is microcontrollers.MessageState$Started@ecc41e
I'm skipping 0
Read state is microcontrollers.MessageState$Started@3681e2
I'm skipping 0
Read state is microcontrollers.MessageState$Started@15e7e8a
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1523a8a
I'm skipping 0
Read state is microcontrollers.MessageState$Started@132d54a
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1537f5f
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1fdf291
I'm skipping 0
Read state is microcontrollers.MessageState$Started@155178e
I'm skipping 0
Read state is microcontrollers.MessageState$Started@48066b
I'm skipping 0
Read state is microcontrollers.MessageState$Started@31343a
I'm skipping 0
Read state is microcontrollers.MessageState$Started@f82157
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1d8bd21
I'm skipping 0
Read state is microcontrollers.MessageState$Started@10793af
I'm skipping 0
Read state is microcontrollers.MessageState$Started@2af851
I'm skipping 0
Read state is microcontrollers.MessageState$Started@18d9b1a
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1241dbf
I'm skipping 0
Read state is microcontrollers.MessageState$Started@7b7ddf
I'm skipping 0
Read state is microcontrollers.MessageState$Started@180ec71
I'm skipping 0
Read state is microcontrollers.MessageState$Started@a8f9d0
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1c8d1df
I'm skipping 0
Read state is microcontrollers.MessageState$Started@cad0fc
I'm skipping 0
Read state is microcontrollers.MessageState$Started@715470
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1724357
I'm skipping 0
Read state is microcontrollers.MessageState$Started@f9288f
I'm skipping 0
Read state is microcontrollers.MessageState$Started@fe48dc
I'm skipping 0
Read state is microcontrollers.MessageState$Started@e25aa7
I'm skipping 0
Read state is microcontrollers.MessageState$Started@139c417
I'm skipping 0
Read state is microcontrollers.MessageState$Started@a132af
I'm skipping 0
Read state is microcontrollers.MessageState$Started@4d3d0f
I'm skipping 0
Read state is microcontrollers.MessageState$Started@7d83
I'm skipping 0
Read state is microcontrollers.MessageState$Started@10c02e8
I'm skipping 0
Read state is microcontrollers.MessageState$Started@cf64ac
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1a982bf
I'm skipping 0
Read state is microcontrollers.MessageState$Started@96156c
I'm skipping 0
Read state is microcontrollers.MessageState$Started@1284894
I'm skipping -2
Read state is microcontrollers.MessageState$Started@8bb6d9
```

The above describes an execution where the "client" sends a receive command (line 2), sees the following sequence of bytes (note that the line wraps), and skips all of them:

```
[-1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 -2]
```

**Comment by whymarrh** [comment]
I'm not really sure how we'd fix this or why this seemingly ends with a "checksum" of `[0 -2]`?

**Comment by arandell93** [comment]
It's interesting that there are only 52 characters in the above message.

A few other things of note:

- `-1` as a signed byte is `0xFF` in hexadecimal (And therefore `-2` is `0xFE`)
- `-86` is indeed `0xAA`. I wanted to confirm.

Do you have any more sessions logged to review?

**Comment by arandell93** [comment]
Reviewing my own code for the [microcontroller](https://github.com/LakeMaps/microcontrollers/blob/master/Arduino/Wireless/WirelessModule_Node1/WirelessModule_Node1.ino) I have noted a few things that have changed/may contribute to the issue that are worth investigating. If I think of/find more I'll edit the comment.

- [Line 29](https://github.com/LakeMaps/microcontrollers/blob/master/Arduino/Wireless/WirelessModule_Node1/WirelessModule_Node1.ino#L29) `const int maxMsgLength = 68;` was shortened from 72. This sets the size of the `sRead` array; [Line 30](https://github.com/LakeMaps/microcontrollers/blob/master/Arduino/Wireless/WirelessModule_Node1/WirelessModule_Node1.ino#L30) `char sRead[maxMsgLength];`. However, the maximum message length [_is_ 68 bytes](https://github.com/LakeMaps/microcontrollers/wiki/Wireless-Communication-Microcontroller-API#command-0x03--wireless-receive).
- [Line 79](https://github.com/LakeMaps/microcontrollers/blob/master/Arduino/Wireless/WirelessModule_Node1/WirelessModule_Node1.ino#L79) `if ((sRead[0] == (char) newMessage) && (0x00 <= sRead[0] <= 0x0F))` apparently I didn't fix this on the wireless module, just [the propulsion](https://github.com/LakeMaps/microcontrollers/commit/6c5f150f58570f4d213ec41aceb7575315137b89#diff-c321fa502e9eced87463276feddf30f9)? Created #13 to resolve this.

**Comment by whymarrh** [comment]
> It's interesting that there are only 52 characters in the above message.

Indeed. I can't really come up with an explanation for why this would happen. If we assume that we've received the trailing part of a message, what would the full message have been? If we assume the first 16 bytes were `0xAA030000000000000000000000000000` the checksum is invalid. I guess it is possible that there was a message in the first 16 bytes. Weird stuff.

**Comment by whymarrh** [comment]
But specifically I can't imagine what would cause the loss of 16 bytes.

**Comment by whymarrh** [comment]
Let's keep comments here specific to this execution and discuss general instability in the tracking issue.

**Comment by arandell93** [comment]
>  If we assume the first 16 bytes were 0xAA030000000000000000000000000000 the checksum is invalid. I guess it is possible that there was a message in the first 16 bytes

Let's assume for a minute that it was the case that we were skipping 16 bytes and the message was valid with data. This would mean we'd have something like `AA 03 01` followed by 13 bytes of valid data which were missed, and then the `FF` and trailing zeros. However, this seems to mean that the RSSI was `00 00` and the checksum was `00 FE`, but we can put that aside for a minute and ask the following:

When hangashore sent its message over the link, does it may sense that the payload only contained 14 bytes of valid data and put in zeros on the trailing end?

**Comment by whymarrh** [comment]
Yes, it's completely possible that the payload contained 14 non-zero bytes followed by zeros.

**Comment by whymarrh** [comment]
A motion value could be `[13, 0, 0, 0, 0, 21, 47, 161, 71, 191]` for example, which is 10 bytes.

**Comment by arandell93** [comment]
> [...] which is 10 bytes

Is a motion value always 10 bytes? Under what circumstances would hangashore send 14 bytes?

**Comment by whymarrh** [comment]
I had meant that the payload could be 14 bytes or fewer, padded with zeros on the end. As it is represented today, a motion value is always 10 bytes and if we're looking to account for the full 13 (or 14) the trailing bytes would be zero.

That said, I don't think this is a particularly useful thought exercise because the execution above was certainly not ***skipping*** the first 16 bytes. The first 16 bytes are MIA. The first control software did not see the first 16 bytes . I don't know where they would have been lost.

**Comment by arandell93** [comment]
> A few other things of note:
> -1 as a signed byte is 0xFF in hexadecimal

We were seeing 0xFF last night as the only byte in reply to a receive command if there was data being sent over the wireless link due to the incorrect wiring of the interrupt line to arduino pin 1 instead of pin 2. This, however, does not explain the `0x00` and `0xFE` that was seen in your OP.

Perhaps we will no longer see this error, however I cannot say for certain that we have corrected the issue as it was seen here. We may have just exposed an existing issue through the use of corrupt data. Thoughts @whymarrh?

**Comment by whymarrh** [comment]
I think this is very likely the issue we solved yesterday evening (refs #14)
